### PR TITLE
Make sure that a valid TCP over DNS message can be decoded

### PR DESCRIPTION
**Pull request by guedou**
This PR partially fixes 2406. It is the root cause of the issue that triggers the recursion loop.

### CODE REVIEW

File: scapy/layers/dns.py [hunk 1]
```diff
@@ -437,6 +437,23 @@ def compress(self):
...
 
+    def pre_dissect(self, s):
+        """
+        Check that a valid DNS over TCP message can be decoded
+        """
+        if isinstance(self.underlayer, TCP):
+            # Compute the length of the DNS packet
+            dns_len = 0
+            if len(s) >= 2:
+                dns_len = struct.unpack("!H", s[:2])[0]
```
Comment (p-l-) [inline scapy/layers/dns.py]: Maybe we should raise the ScapyException error here too?

Comment (guedou) [reply scapy/layers/dns.py]: You mean to explicitly raise an exception if `len(s) < 2` ?

Comment (p-l-) [reply scapy/layers/dns.py]: Yep. I think it would make sense.

Comment (guedou) [reply scapy/layers/dns.py]: Done.

**Pull request merged**